torch/lib/THCS/generic/THCSTensor.cu

```diff
@@ -14,68 +30,178 @@ THCTensor *THCSTensor_(toDense)(THCState *state, THCSTensor *self) {
+  } else {
+    THArgCheck(getApplyGrid(state, nnz * block.x, grid), 1, CUTORCH_DIM_WARNING);
 
-  THCSTensor_toDenseKernel<unsigned long, real>
-    <<<grid, block, 0, THCState_getCurrentStream(state)>>>(
-        otherInfo, indicesInfo, valuesInfo, (unsigned long)(nnz));
+    THCSTensor_spcKernel<TensorAddOp<real>, unsigned long, real>
+      <<<grid, block, 0, THCState_getCurrentStream(state)>>>(
+          TensorAddOp<real>(),
+          V_INFO(other), I_INFO(self->indices), V_INFO(self->values),
+          (unsigned long)(nnz));
+  }
 
   THCudaCheck(cudaGetLastError());
   THLongStorage_free(storage);
   return other;
-  */
-  THError("WARNING: Sparse Cuda Tensor op toDense is not implemented");
-  return NULL;
 }
 
 void THCSTensor_(reorder)(THCState *state, THCSTensor *self) {
-  THError("WARNING: Sparse Cuda Tensor op reorder is not implemented");
-}
+  if (self->nnz < 2) return;
+#if CUDA_VERSION >= 7000
+  THCThrustAllocator thrustAlloc(state);
+#define THRUST_EXEC(fn, ...) fn(thrust::cuda::par(thrustAlloc).on(THCState_getCurrentStream(state)), ##__VA_ARGS__)
+#else
+#define THRUST_EXEC(fn, ...) fn(##__VA_ARGS__)
+#endif
+
+  THCIndexTensor *indices = THCSTensor_(indices)(state, self);
+  THCTensor *values = THCSTensor_(values)(state, self);
+  THCIndexTensor *indicesSlice = THCIndexTensor_(new)(state);
+  THCIndexTensor *indicesScalar = THCIndexTensor_(newWithSize1d)(state, self->nnz);
+  THCIndexTensor *projectIndices = THCIndexTensor_(newWithSize2d)(state, 2, self->nnz);
+  THCTensor *projectValues = THCTensor_(newWithSize1d)(state, self->nnz);
+  THCTensor_(fill)(state, projectValues, ScalarConvert<int, real>::to(1));
+  THCudaLongTensor *mapping = THCudaLongTensor_new(state);
+  THCudaLongTensor *permutation = THCudaLongTensor_new(state);
+  THCudaLongTensor_select(state, mapping, projectIndices, 0, 0);
+  THCudaLongTensor_select(state, permutation, projectIndices, 0, 1);
+  THCudaLongTensor *unique = THCudaLongTensor_newWithSize1d(state, self->nnz);
+
+  thrust::device_ptr<long> permutationIter(THCudaLongTensor_data(state, permutation));
+  THRUST_EXEC(thrust::sequence, permutationIter, permutationIter + self->nnz);
+  THCIndexTensor_(zero)(state, indicesScalar);
+  integer factor = 1;
+  for (int i = self->nDimensionI - 1; i >= 0; i--) {
+    THCIndexTensor_(select)(state, indicesSlice, indices, 0, i);
+    THCIndexTensor_(cadd)(state, indicesScalar, indicesScalar, factor, indicesSlice);
+    factor *= self->size[i];
+  }
+  thrust::device_ptr<integer> indicesIter(THCIndexTensor_(data)(state, indicesScalar));
+
+  // HACK *theoretically* we need stable sort, so that indices mapped to the same
+  // location are sorted in the permutation. This is necessary to ensure that
+  // the projection matrix is contiguous
+  // However, it seems to work even with non-stable sort, which is nice because it's faster...
+  // Revert this if bugs arise, or add a step to sort the projection matrix indices explicitly
```

Comment:
It works likely because most (if not all) of the implementations in THCTensor_(sort) also do stable sort. Fastest algo on gpu generally is radix, which also happens to be stable.